pmlc/dialect/pxa/transforms/stencil_generic.cc

```diff
@@ -0,0 +1,164 @@
+      new_bound_idxs.push_back(block_arg);
+      RecursiveBindIndex(new_bound_idxs, tensors);
+    }
+  }
+}
+
+void StencilGeneric::DoStenciling() {
+  // Initialization
+  // TODO: Something like?: tensors.clear();
+  {
+    auto maybe_loads_and_stores = capture();
+    if (maybe_loads_and_stores) {
+      loads_and_stores = maybe_loads_and_stores.getValue();
+    } else {
+      IVLOG(4, "Cannot Stencil: Operations fail to pattern-match.");
+      return;
+    }
+  }
+
+  llvm::SmallVector<mlir::Value, 3> tensors;
+  for (auto &load_op : loads_and_stores.loads) {
+    // TODO: It's probably fine, but I'm a little worried `getMemRef` isn't
+    // const
+    tensors.push_back(load_op.getMemRef());
+  }
+  size_t first_store_idx = tensors.size();
+  for (auto &store_op : loads_and_stores.stores) {
+    // TODO: It's probably fine, but I'm a little worried `out` isn't const
+    tensors.push_back(store_op.out());
+    // TODO: Probably should handle reduces vs. true stores in a different way
+    // if (auto reduce_op = llvm::dyn_cast_or_null<AffineReduceOp>(store_op)) {
+    //   tensors.push_back(reduce_op.out());
+    // } else if (auto true_store_op =
+    // llvm::dyn_cast_or_null<mlir::AffineStoreOp>(store_op)) {
+    //   tensors.push_back(true_store_op.getMemRef());
+    // } else {
+    //   // TODO: throw?
+    //   IVLOG(1, "Unexpected failure to load tensors from ops in stenciling");
+    //   return;
+    // }
+  }
+  std::sort(tensors.begin(), tensors.begin() + first_store_idx);
+  do { // Each load tensor permutation
+    std::sort(tensors.begin() + first_store_idx, tensors.end());
+    do { // Each store tensor permutation
+      // Add all legal permutations to legal_permutations
+      RecursiveBindIndex(llvm::SmallVector<mlir::BlockArgument, 8>(), tensors);
+    } while (std::next_permutation(tensors.begin() + first_store_idx,
+                                   tensors.end()));
+  } while (std::next_permutation(tensors.begin(),
+                                 tensors.begin() + first_store_idx));
+
+  // TODO: If we desire more complex requirements than pairwise tensor-to-index
+  // stride requirements, that function could go here
+
+  for (const auto &perm : legal_permutations) {
+    // TODO: Clean, try to get everything on the same unsigned type/width
+    auto ranges = op.getConstantRanges();
+    if (!ranges) {
+      // Can't stencil without constant ranges TODO: Check this earlier, not in
+      // the middle of some late loop!
+      IVLOG(1, "NO CONSTANT RANGES TODO");
+      break;
+    }
+    for (size_t i = 0; i < perm.indexes.size(); i++) {
+      llvm::SmallVector<size_t, 8> tile_size;
+      size_t idx_in_block = perm.indexes[i].getArgNumber();
+      try { // TODO: probably don't keep long term
+        for (const auto &size :
+             tiling_generators[i](ranges.getValue()[idx_in_block])) {
+          tile_size.push_back(size);
+        }
+      } catch (const std::bad_function_call &e) {
```

Comment:
What is this about?

Reply:
WIP code -- if sufficient `tiling_generators` aren't initialized that's the error that comes up. In finalized form, I suspect we want to test early that the requirements/tiling_generators are of consistent sizes and not do a `try` block here?

Reply:
We should try not to use exceptions at all. The MLIR way is to return errors. If we are introducing new exceptions, then we must have a catch just before the pass returns control to MLIR. MLIR cannot handle exceptions (in fact they disable exceptions in upstream builds).

Reply:
Are asserts also reasonable in MLIR land? This is a situation that should never happen (it comes from calling an uninitialized `std::function`)